### ISSUE TITLE
mat: Rename Clone to CloneFrom and add Clone methods.

### ISSUE THREAD

**Issue by btracey** [gonum/gonum#991]
There's currently no way to get the clone of the matrix with the type preserved. The obvious signature for this is  `func (m *DenseFoo) Clone() Matrix`. Unfortunately, this clashes with the current `func (m *Dense) Clone(a Matrix)`. I propose that this become `func (m *Dense) CloneFrom(a Matrix)`, which matches `DiagFrom` and the other `From` types. If we don't want to change the method, we can use `func (m *DenseFoo) CloneOf() Matrix`, so it's not the only way.

**Comment by kortschak** [comment]
I agree with the `CloneFrom` change; `CloneOf` doesn't work properly when you read it.

**Comment by kortschak** [comment]
@btracey For this, are you also thinking that `CloneVec() Vector`, `CloneTri() Triangular` etc would also exist?
